test/src/unit-max-tile-size.cc

```diff
@@ -0,0 +1,1149 @@
+  };
+
+  template <const int nchars = 1>
+  void write_array_attr(const std::string& array_name, const std::string&& attr_name) {
+    tiledb::Context ctx;
+
+    // Prepare some data for the array
+    std::vector<std::array<char, nchars>> a;
+    for (auto i = 0; i < 4; ++i) {
+      std::array<char, nchars> buf;
+      memset(&buf[0], 'a' + i, nchars);
+      a.emplace_back(buf);
+    }
+
+    // Open the array for writing and create the query.
+    tiledb::Array array(ctx, array_name, TILEDB_WRITE);
+    tiledb::Query query(ctx, array);
+    query.set_layout(TILEDB_ROW_MAJOR).set_data_buffer(attr_name, a);
+
+    // Perform the write and close the array.
+    query.submit();
+    array.close();
+  };
+
+  uint64_t c_get_fragments_max_in_memory_tile_size(const std::string& array_uri) {
+    tiledb::Context ctx;
+    uint64_t max_in_memory_tile_size = 0;
+    CHECK(
+        tiledb_array_maximum_tile_size(
+            &*ctx.ptr(),
+            array_uri.c_str(),
+            &max_in_memory_tile_size, 
+            nullptr) == TILEDB_OK);
+
+    if (showing_data) {
+      std::cout << "maximum in memory tile size "
+                << max_in_memory_tile_size
+                << std::endl;
+    }
+    return max_in_memory_tile_size;
+  }
+
+  uint64_t cpp_get_fragments_max_in_memory_tile_size(const std::string& array_uri) {
+    tiledb::Context ctx;
+    tiledb::Array array(ctx, array_uri, TILEDB_READ);
+    uint64_t max_in_memory_tile_size = 0;
+    array.get_max_in_memory_tile_size(&max_in_memory_tile_size);
+
+    if (showing_data) {
+      std::cout << "maximum in memory tile size " << max_in_memory_tile_size
+                << std::endl;
+    }
+    return max_in_memory_tile_size;
+  }
+
+  tiledb::Context ctx_;
+  tiledb::VFS vfs_;
+};
+
+TEST_CASE_METHOD(
+    CandCXXAPI_MaxTileSizeFx,
+    "C++ API: Max tile size, dense, fixed, with consolidation",
+    "[capi][cppapi][max-tile-size][dense][consolidate]") {
+  // Name of array.
+  auto& array_name = main_array_name;
+
+  auto create_array = [&]() -> void {
+    // Create a TileDB context.
+    tiledb::Context ctx;
+
+    remove_temp_dir(array_name.c_str());
+
+    // The array will be 4x4 with dimensions "rows" and "cols", with domain
+    // [1,4] and space tiles 2x2
+    tiledb::Domain domain(ctx);
+    domain
+        .add_dimension(tiledb::Dimension::create<int>(ctx, "rows", {{1, 4}}, 2))
+        .add_dimension(
+            tiledb::Dimension::create<int>(ctx, "cols", {{1, 4}}, 2));
+
+    // The array will be dense.
+    tiledb::ArraySchema schema(ctx, TILEDB_DENSE);
+    schema.set_domain(domain).set_order({{TILEDB_ROW_MAJOR, TILEDB_ROW_MAJOR}});
+
+    // Add a single attribute "a" so each (i,j) cell can store an integer.
+    schema.add_attribute(tiledb::Attribute::create<int32_t>(ctx, "a"));
+
+    // Create the (empty) array on disk.
+    tiledb::Array::create(array_name, schema);
+  };
+
+  auto write_array_1 = [&]() -> void {
```

Comment:
I really think these need to be moved to a helper of  CandCXXAPI_MaxTileSizeFx. The 4 functions are doing exactly the same thing. If you make a helper like:
write_array_2_dim(std::vector<int> data, std::array<int, 2> dim1_ranges, std::array<int, 2> dim2_ranges), this will shorten the code quite a bit and also allow reuse across test cases. This will also make the test cases much shorter in length and easier to review.

For this API, I think you could have helpers to create and write to arrays with 3-4 different schemas. Then the test cases themselves would be short and it's easier to see the test coverage as a whole. There is also an argument that can be made for consistency with the rest of the test files in for the API tests.

Reply:
changes made